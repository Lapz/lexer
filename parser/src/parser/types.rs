

use crate::parser::Parser;

use crate::{Span, SyntaxKind::*, Token};

impl<'a, I> Parser<'a, I>
where
    I: Iterator<Item = Span<Token>>,
{
    pub(crate) fn parse_type(&mut self) {
        match self.current() {
            IDENT | T![void] => self.parse_ident_type(),
            T!["["] => self.parse_array_type(),
            T!["("] => self.parse_paren_type(),
            T![fn] => self.parse_fn_type(),
            e => {
                println!("{:?}", e);
                self.error("Expected a type parameter  ")
            }
        };
    }

    fn parse_ident_type(&mut self) {
        self.start_node(IDENT_TYPE);

        if self.matches(vec![T![nil], IDENT]) {
            self.bump();
        } else {
            self.error("Expected an identifier or `nil`")
        }

        if self.is_ahead(|t| t == T![<]) {
            self.parse_type_params(false);
        }

        self.finish_node();
    }

    fn parse_array_type(&mut self) {
        self.start_node(ARRAY_TYPE);
        self.bump(); //Eat `[`

        self.parse_type();

        self.expect(T!["]"], "Expected `]`");

        self.finish_node();
    }

    fn parse_paren_type(&mut self) {
        self.start_node(PAREN_TYPE);
        self.bump(); //Eat `(`

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"], "Expected `)`");

        self.finish_node();
    }

    fn parse_fn_type(&mut self) {
        self.start_node(FN_TYPE);
        self.bump(); //Eat `fn`

        self.expect(T!["("], "Expected `(`");

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"], "Expected `)`");

        if self.is_ahead(|t| t == T![->]) {
            self.bump();

            self.parse_type();
        }

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_fn_type,"fn main(_:fn(i32,i32) -> i32) {}"}
    test_parser! {parse_fn_tuple_type,"fn main(_:fn((i32,i32)) -> i32) {}"}
    test_parser! {parse_ident_type,"fn main(x:i32) {}"}
    test_parser! {parse_array_type,"fn main(x:[i32]) {}"}
    test_parser! {parse_array_tuple_type,"fn main(x:[(i32,i32)]) {}"}
    test_parser! {parse_tuple_type,"fn main(x:(i32,i32)) {}"}
}
