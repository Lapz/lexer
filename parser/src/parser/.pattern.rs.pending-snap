{"run_id":"a9ba4d58-8676-410f-9d7e-fe088fc2dc51","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:24:41.934957Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"SOURCE_FILE@[0; 13)\\n  ITEM_LIST@[0; 13)\\n    FN_DEF@[0; 13)\\n      FN_KW@[0; 2) \\\"fn\\\"\\n      IDENT@[2; 6) \\\"main\\\"\\n      PARAM_LIST@[6; 13)\\n        L_PAREN@[6; 7) \\\"(\\\"\\n        UNDERSCORE@[7; 8) \\\"_\\\"\\n        COLON@[8; 9) \\\":\\\"\\n        IDENT@[9; 12) \\\"i32\\\"\\n        R_PAREN@[12; 13) \\\")\\\"\\n\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"\n        PLACEHOLDER_PAT\n            UNDERSCORE\n        "}}
{"run_id":"cb33af26-1bf3-4421-9cd5-8b7cc9951033","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:25:37.802132Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"SOURCE_FILE@[0; 13)\\n  ITEM_LIST@[0; 13)\\n    FN_DEF@[0; 13)\\n      FN_KW@[0; 2) \\\"fn\\\"\\n      IDENT@[2; 6) \\\"main\\\"\\n      PARAM_LIST@[6; 13)\\n        L_PAREN@[6; 7) \\\"(\\\"\\n        UNDERSCORE@[7; 8) \\\"_\\\"\\n        COLON@[8; 9) \\\":\\\"\\n        IDENT@[9; 12) \\\"i32\\\"\\n        R_PAREN@[12; 13) \\\")\\\"\\n\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"\n        PLACEHOLDER_PAT\n            UNDERSCORE\n        "}}
{"run_id":"fa90cf48-8844-4ae2-b2f7-6a1440c5fa2b","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:27:50.329174Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"SOURCE_FILE@[0; 13)\\n  ITEM_LIST@[0; 13)\\n    FN_DEF@[0; 13)\\n      FN_KW@[0; 2) \\\"fn\\\"\\n      IDENT@[2; 6) \\\"main\\\"\\n      PARAM_LIST@[6; 13)\\n        L_PAREN@[6; 7) \\\"(\\\"\\n        UNDERSCORE@[7; 8) \\\"_\\\"\\n        COLON@[8; 9) \\\":\\\"\\n        IDENT@[9; 12) \\\"i32\\\"\\n        R_PAREN@[12; 13) \\\")\\\"\\n\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"\n        PLACEHOLDER_PAT\n            UNDERSCORE\n        "}}
{"run_id":"90097682-43df-47ab-b00f-634de650687b","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:39:26.972007Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"SOURCE_FILE@[0; 13)\\n  ITEM_LIST@[0; 13)\\n    FN_DEF@[0; 13)\\n      FN_KW@[0; 2) \\\"fn\\\"\\n      IDENT@[2; 6) \\\"main\\\"\\n      PARAM_LIST@[6; 13)\\n        L_PAREN@[6; 7) \\\"(\\\"\\n        UNDERSCORE@[7; 8) \\\"_\\\"\\n        COLON@[8; 9) \\\":\\\"\\n        IDENT@[9; 12) \\\"i32\\\"\\n        R_PAREN@[12; 13) \\\")\\\"\\n\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
{"run_id":"cf33fcf7-cc99-412e-8809-7152c5768d1a","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:46:37.928197Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"\\\"fnmain(_:i32)\\\"\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
{"run_id":"00060ec3-d391-42c4-8b90-fecf59bf4dc3","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:50:08.681971Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"\\\"fnmain(_:i32)\\\"\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
{"run_id":"5efd478a-eb0f-463a-a9eb-fa6ec1d75af6","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:50:37.893632Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"SOURCE_FILE@[0; 13)\\n  ITEM_LIST@[0; 13)\\n    FN_DEF@[0; 13)\\n      FN_KW@[0; 2) \\\"fn\\\"\\n      IDENT@[2; 6) \\\"main\\\"\\n      PARAM_LIST@[6; 13)\\n        L_PAREN@[6; 7) \\\"(\\\"\\n        UNDERSCORE@[7; 8) \\\"_\\\"\\n        COLON@[8; 9) \\\":\\\"\\n        IDENT@[9; 12) \\\"i32\\\"\\n        R_PAREN@[12; 13) \\\")\\\"\\n\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
{"run_id":"2024b988-4a14-4763-a48a-c921f7be9204","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:51:00.885266Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"fnmain(_:i32)\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
{"run_id":"c9471ebd-0551-43f0-b02e-9b230e25cf5e","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:51:09.321732Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"fnmain(_:i32)\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
{"run_id":"88ec945a-40fc-495f-9c67-50499a5ca724","line":71,"new":{"module_name":"test","metadata":{"created":"2019-08-08T20:56:54.372957Z","creator":"insta@0.8.2","source":"parser/src/parser/pattern.rs","expression":"dump_debug(&parser_output)"},"snapshot":"\"SOURCE_FILE@[0; 13)\\n  ITEM_LIST@[0; 13)\\n    FN_DEF@[0; 13)\\n      FN_KW@[0; 2) \\\"fn\\\"\\n      IDENT@[2; 6) \\\"main\\\"\\n      PARAM_LIST@[6; 13)\\n        L_PAREN@[6; 7) \\\"(\\\"\\n        UNDERSCORE@[7; 8) \\\"_\\\"\\n        COLON@[8; 9) \\\":\\\"\\n        IDENT@[9; 12) \\\"i32\\\"\\n        R_PAREN@[12; 13) \\\")\\\"\\n\""},"old":{"module_name":"test","metadata":{"created":"2019-08-08T07:55:29.701146023Z"},"snapshot":"SOURCE_FILE@[0; 13)\n        ITEM_LIST@[0; 13)\n        "}}
