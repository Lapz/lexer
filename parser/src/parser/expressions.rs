mod block;
mod break_expr;
mod continue_expr;
