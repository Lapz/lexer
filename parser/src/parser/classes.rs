use syntax::T;

use crate::parser::Parser;

use crate::{Span, SyntaxKind::*, Token};

impl<'a, I> Parser<'a, I>
where
    I: Iterator<Item = Span<Token>>,
{
    pub(crate) fn parse_class(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, CLASS_DEF);

        self.expect(CLASS_KW, "Expected `class`");

        self.ident();

        if self.is_ahead(|t| t == T![<]) {
            self.parse_type_params(false);
        }

        self.parse_class_body();

        self.finish_node()
    }

    fn parse_class_body(&mut self) {
        // self.start_node()
        self.expect(T!["{"], "Expected `{`");

        while !self.at(EOF) && !self.at(T!["}"]) {
            let has_visibility = self.has_visibility();

            if has_visibility {
                let checkpoint = self.checkpoint();
                self.parse_visibility();

                match self.current() {
                    IDENT => self.parse_named_field(),
                    T![fn] => self.parse_function(checkpoint),
                    T![] => {
                        self.bump();
                        continue;
                    }
                    _ => self.error(
                        "Expected an identifier | `pub` | `fn` ",
                        format!(
                            "Expected an identifier | `pub` | `fn`  but instead found `{}`",
                            self.current_string()
                        ),
                    ),
                }
            } else {
                let checkpoint = self.checkpoint();
                match self.current() {
                    IDENT => self.parse_named_field(),
                    T![fn] => self.parse_function(checkpoint),
                    T![] => {
                        self.bump();
                        continue;
                    }
                    _ => self.error(
                        "Expected an identifier | `pub` | `fn` ",
                        format!(
                            "Expected an identifier | `pub` | `fn`  but instead found `{}`",
                            self.current_string()
                        ),
                    ),
                }
            }
        }

        self.expect(T!["}"], "Expected `}`");
    }

    fn parse_named_field(&mut self) {
        self.start_node(NAMED_FIELD_DEF);
        self.ident();
        self.expect(T![:], "Expected `:`");
        self.parse_type();
        self.expect(T![;], "Expected `;`");
        self.finish_node();
    }
}

#[cfg(test)]
mod test {

    use crate::utils::parse;
    use syntax::{ClassDefOwner, NamedFieldsOwner};
    #[test]
    fn test_class_fields() {
        let source_file = parse("class Person { name:String; surname:String;}").parse_program();

        let class = source_file.classes().next().unwrap();

        assert_eq!(class.fields().count(), 2)
    }
    test_parser! {parse_empty_class,"class Foo {}"}
    test_parser! {parse_class_generic,"class Result<T,E> {}"}
    test_parser! {parse_class_fields,"class Person { name:String; surname:String;}"}
    test_parser! {parse_class_fields_methods,"class Person { name:String; surname:String; fn hello(self) {}}"}
    test_parser! {parse_class_methods,"class Person { name:String; surname:String; fn new() -> Person {}}"}
}
