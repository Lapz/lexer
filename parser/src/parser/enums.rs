use crate::T;

use crate::parser::Parser;

use crate::{Span, SyntaxKind::*, Token};

impl<'a, I> Parser<'a, I>
where
    I: Iterator<Item = Span<Token>>,
{
    pub(crate) fn parse_enum(&mut self, has_visibility: bool) {
        self.start_node(ENUM_DEF);

        if has_visibility {
            self.parse_visibility();
        }

        self.expect(T![enum], "Expected `enum`");

        self.ident();

        if self.is_ahead(|t| t == T![<]) {
            self.parse_type_params(false);
        }

        self.parse_enum_variants();

        self.finish_node()
    }

    fn parse_enum_variants(&mut self) {
        self.start_node(ENUM_VARIANT_LIST);

        self.expect(T!["{"], "Expected `{`");

        while !self.at(EOF) && !self.at(T!["}"]) {
            self.parse_enum_variant();
            if !self.at(T!["}"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T!["}"], "Expected `}`");

        self.finish_node();
    }

    fn parse_enum_variant(&mut self) {
        self.start_node(ENUM_VARIANT);

        self.ident();

        if self.is_ahead(|t| t == T!["("]) {
            self.parse_enum_variant_types()
        }

        self.finish_node();
    }

    fn parse_enum_variant_types(&mut self) {
        self.bump(); // Eat the "("
        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_type();

            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"], "Expected `>` to close type params");
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_empty_enum,"enum Foo {}"}
    test_parser! {parse_enum_with_variants,"enum Foo {A,B,C}"}
    test_parser! {parse_enum_variants_trailing_comma,"enum Foo {A,B,C,}"}
    test_parser! {parse_enum_variants_types,"enum Foo { A([i32]),B(i32),C((i32,i32)),}"}
    test_parser! {parse_enum_generic,"enum Result<T,E> { Ok(T),Err(E)}"}
}
