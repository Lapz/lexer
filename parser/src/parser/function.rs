use syntax::T;

use crate::parser::Parser;

use crate::{Span, SyntaxKind::*, Token};

impl<'a, I> Parser<'a, I>
where
    I: Iterator<Item = Span<Token>>,
{
    pub(crate) fn parse_function(&mut self, checkpoint: rowan::Checkpoint) {
        self.start_node_at(checkpoint, FN_DEF);

        self.expect(T![fn], "Expected `fn`");

        self.ident();

        if self.current() == T![<] {
            self.parse_type_params(false);
        }

        if self.current() == T!["("] {
            self.parse_func_params();
        }

        if self.current() == T![->] {
            self.parse_return_type();
        }

        self.parse_block();

        self.finish_node()
    }

    fn parse_return_type(&mut self) {
        self.start_node(RET_TYPE);
        self.expect(T![->], "Expected `->`");
        self.parse_type();
        self.finish_node()
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_function,"fn main() {}"}
}
