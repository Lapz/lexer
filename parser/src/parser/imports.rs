use syntax::T;

use crate::parser::Parser;

use crate::SyntaxKind::*;

impl<'a> Parser<'a> {
    pub(crate) fn parse_import(&mut self) {
        self.start_node(IMPORT_DEF);

        self.expect(T![import]);

        self.parse_import_segment();

        self.expect(T![;]);

        self.finish_node();
    }

    pub(crate) fn parse_import_segment(&mut self) {
        self.start_node(IMPORT_SEGMENT);

        self.ident();

        if self.at(T![::]) {
            self.bump();

            if self.at(T!["{"]) {
                self.bump();

                while !self.at(EOF) && self.at(T!["}"]) {
                    self.parse_import_segment();

                    if !self.at(T!["}"]) && !self.expected(T![,]) {
                        break;
                    }
                }

                self.expect(T!["}"]);
            }

            self.parse_import_segment();
        }

        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_empty_import,"import foo;"}
    test_parser! {parse_import_with_multiple_segments,"import foo::bar::baz;"}
    test_parser! {parse_import_with_brace,"import foo::{bar,baz};"}
    test_parser! {parse_import_with_brace_nested_segments,"import foo::{bar::baz,fubar};"}
}
