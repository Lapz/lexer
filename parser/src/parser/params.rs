use crate::parser::Parser;
use crate::T;

use crate::{Span, SyntaxKind::*, Token};

impl<'a, I> Parser<'a, I>
where
    I: Iterator<Item = Span<Token>>,
{
    pub(crate) fn parse_func_params(&mut self) {
        self.start_node(PARAM_LIST);
        self.bump();

        while !self.at(EOF) && !self.at(T![")"]) {
            self.func_param();

            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"], "Expected `)` to close type params");
        self.finish_node()
    }

    fn func_param(&mut self) {
        self.start_node(PARAM);
        self.parse_pattern(false);
        self.expect(COLON, "Expected `:`");
        self.parse_type();
        self.finish_node();
    }
}
