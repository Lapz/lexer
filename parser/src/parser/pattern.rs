use crate::parser::Parser;
use crate::T;

use crate::{Span, SyntaxKind::*, Token};

impl<'a, I> Parser<'a, I>
where
    I: Iterator<Item = Span<Token>>,
{
    pub(crate) fn parse_pattern(&mut self, allow_literal: bool) {
        match self.current() {
            T!["("] => self.parse_tuple_pattern(allow_literal),
            IDENT => self.parse_binding_pattern(),
            T![_] => self.parse_placeholder_pattern(),
            _ => {
                if allow_literal {
                    // self.parse_literal()
                } else {
                    self.error(
                        "Expected a literal pattern",
                        format!(
                            "Expected a literal pattern instead found `{}`",
                            self.current_string()
                        ),
                    )
                }
            }
        }
    }

    fn parse_tuple_pattern(&mut self, allow_literal: bool) {
        self.start_node(TUPLE_PAT);

        self.bump(); // Eat the `(`

        while !self.at(EOF) && !self.at(T![")"]) {
            self.parse_pattern(allow_literal);
            if !self.at(T![")"]) && !self.expected(T![,]) {
                break;
            }
        }

        self.expect(T![")"], "Expected `)` to close type params");
        self.finish_node();
    }

    fn parse_placeholder_pattern(&mut self) {
        self.start_node(PLACEHOLDER_PAT);
        self.expect(T![_], "Expected `_`");
        self.finish_node();
    }

    fn parse_binding_pattern(&mut self) {
        self.start_node(BIND_PAT);
        self.ident();
        self.finish_node();
    }
}

#[cfg(test)]
mod test {
    test_parser! {parse_placeholder_pattern,"fn main(_:i32) {}"}

    test_parser! {parse_tuple_pattern,"fn main((x,y):i32) {}"}

    test_parser! {parse_binding_pattern,"fn main(x:i32) {}"}

    test_parser! {parse_nested_tuple_pattern,"fn main((x,(y,_)):i32) {}"}
}
