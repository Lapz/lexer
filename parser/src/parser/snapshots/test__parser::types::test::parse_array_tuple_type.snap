---
created: "2019-08-11T21:50:37.801603Z"
creator: insta@0.8.2
source: parser/src/parser/types.rs
expression: "crate::utils::dump_debug(&parser_output)"
---
SOURCE_FILE@[0; 23)
  FN_DEF@[0; 21)
    FN_KW@[0; 2) "fn"
    NAME@[2; 6)
      IDENT@[2; 6) "main"
    PARAM_LIST@[6; 21)
      L_PAREN@[6; 7) "("
      PARAM@[7; 20)
        PLACEHOLDER_PAT@[7; 8)
          NAME@[7; 8)
            IDENT@[7; 8) "x"
        COLON@[8; 9) ":"
        ARRAY_TYPE@[9; 20)
          L_BRACK@[9; 10) "["
          PAREN_TYPE@[10; 19)
            L_PAREN@[10; 11) "("
            IDENT_TYPE@[11; 14)
              IDENT@[11; 14) "i32"
            COMMA@[14; 15) ","
            IDENT_TYPE@[15; 18)
              IDENT@[15; 18) "i32"
            R_PAREN@[18; 19) ")"
          R_BRACK@[19; 20) "]"
      R_PAREN@[20; 21) ")"
  L_CURLY@[21; 22) "{"
  R_CURLY@[22; 23) "}"

